### ISSUE TITLE
Shape of color temperature curve

### ISSUE THREAD

**Issue by jarpatus** [basnijholt/adaptive-lighting#62]
If I got this right then color temperature curve used by this component looks like half sine wave (max. color temperature reacher only for a short period of time). Any change this could be controlled somehow? I would like color temperature curve to look more like brightness curve looks like (kind of square wave with smoothened rise and fall). Perhaps adding optional configuration options like rise time and fall time?

For background info; I am ATM using flux component and hacked functionality like this into it (actually I set dawn, sunrise, sunset and dusk times and color temperatures and color temperature will only change between dawn-sunrise and sunset-dusk). I have found this to be superior compared to out-of-the-box flux and if I got this right out-of-the-box adaptive-lighting. I live in northern countries and during winter I get very little real sunlight so this may affect on one's experience what is good curve and what is not. 

I thought about creating pull request for my flux modifications but looks like this adaptive-lighting is superior in every way so ditched the idea...

**Comment by basnijholt** [comment]
What you suggest seems reasonable. Could you share your flux edits somewhere such that I can take a look at the code?

> Perhaps adding optional configuration options like rise time and fall time

This isn't possible now, but I will consider doing this.

**Comment by jarpatus** [comment]
My customized flux can be found from:

https://github.com/jarpatus/core/blob/dev/homeassistant/components/flux/switch.py

Disclaimer: I am no python developer nor home assistant developer, so take this with grain of salt.

**Comment by jd1900** [comment]
This would be great.
Maybe set a time frame in which temperature is not affected (for instance from sunrise to 1 hour before sunset).
Afterall, talking about circadian rithms, once sun has risen, it doesn't make sense to don't have a cold/blue light once sun has risen. It doesn't have to be gradually increased.

My proposal would be to have a few time ranges where you can set the start and end time.
- TRANSIENT TO DAY LIGHT. From warm to cold. For instance, you can set to zero this transient to day light, thus making it instant, or define it.
- DAY LIGHT. During this period colour temperature and brightness remain equal, no changes are made.
- TRANSIENT TO DAY LIGHT: Same as transient to day light but from cold to warm.
- NIGHT LIGHT: Same as day light. During this period colour temperature and brightness remain equal, no changes are made.

**Comment by euinor** [comment]
More control over the color and temperature curves would be great.
My use case is that in winter I _don't_ want to follow sunrise/sunset. I would rather my circadian rhythm followed a standard working day (100% blue and bright 9am till 5.30) even though it gets dark before 5.30.

**Comment by RubenKelevra** [comment]
Hey @jarpatus, @jd1900, @euinor,

I feel like that's a duplicate of #100 and #131 - do you agree?

**Comment by jarpatus** [comment]
@RubenKelevra at least #100, if implemented, would be superset of this one (if you can manually set color temperatures for each start times so you can shape your curve as you wish). So yeah I agree!

**Comment by RubenKelevra** [comment]
> @RubenKelevra at least #100, if implemented, would be superset of this one (if you can manually set color temperatures for each start times so you can shape your curve as you wish). So yeah I agree!

Okay, no that's not the case.

I feel like #100 only covers that we have like a natural and a synthetic curve for the color temperature. :)

**Comment by RubenKelevra** [comment]
@sstarcher wrote:
> I'm in a similar position where I want the morning light to not really follow the sun and to instead lock my circadian rhythm in to help wake me up before work.

Well, that's a time shift of the sunrise, which you can do right now. Just go to the settings and select a different time for the sunrise.

> This very much interests me and if we can define an acceptable way to implement this I would be happy to add this feature.

Well, while I can see that some might think that this is what's the solution I doubt that there are many people willing to either create hundreds of points on a LUT or have the formula to do the same to define their own color temperatures.

I think what people really want is that the current functionality works better, closer to reality.

While custom color curves might be a feature in the future I think it's best to keep this to PRs to the project, to ship the work to other users as well.

Additionally many users might just have different lights, so their perceptions of how the program is currently working is different. That's why it's important to have color temperature as well brightness corrections (https://github.com/basnijholt/adaptive-lighting/issues/124).

**Comment by RubenKelevra** [comment]
It makes sense to reexamine this decision after version 2.0 release

**Comment by sstarcher** [comment]
And to be clear on my point.  I don't want "sunrise".  I want 100% or near 100% brightness first thing in the morning and a lowering to sunset in the evening.  For helping you wake up you are better with extra brightness in the morning to set your clock.

**Comment by Soukyuu** [comment]
Was about to create an issue for this. At the moment, with max kelvin set to 4000 (I don't want the light to get any cooler), the lights are too warm at 6:30am compared to the sunlight. Ideally, I'd like them to either match or be cooler than sunlight for those murky winter days to come. From what I see, customizing the shape of the color and brightness curves would allow me to do just that.

**Comment by danishru** [comment]
+1
Apple HomeKit adaptive lighting works in a similar way as discussed in this thread. In another issue, I described in more detail my observation of Apple HomeKit adaptive lighting, and also attached a curve graph and text data.

**Comment by smaddock** [comment]
The brightness graph has a few different algorithms one can choose from (default, linear, tanh). If the same options were implemented for color temperature, that seems like it would fulfill this request. (I came here assuming that it did, as I want to change the color temp only during my artificial sunrise/sunset. I can currently do this with the native Hue automations, but I just started migrating everything to HA.)

**Comment by emanuelbaltaretu** [comment]
> The brightness graph has a few different algorithms one can choose from (default, linear, tanh). If the same options were implemented for color temperature, that seems like it would fulfill this request. (I came here assuming that it did, as I want to change the color temp only during my artificial sunrise/sunset. I can currently do this with the native Hue automations, but I just started migrating everything to HA.)

Yep, came here to open an issue believing the the brightness graphs will modify temperature as well and both be in synch and tough I found a bug. Great idea with the Hue automations, I will disable adaptive lighting as well in the meantime and use hue bridge as the only thing i want is for temperature and color to decrease/shift to warm from 5PM till 10PM. 
Currently with adaptive lighting at 5PM temperature is already half warm and I like to extend the day in winter times

**Comment by RubenKelevra** [comment]
I'm experimenting with natural light mirroring at the moment. So here's how natural light looks like vs the color curve of adaptive lighting:

![Screenshot_2025-02-24-21-46-34-570-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/1c6fbd96-689b-4c25-ae8b-640a2d5fe6ca)

**Comment by RubenKelevra** [comment]
@Soukyuu wrote:
> Was about to create an issue for this. At the moment, with max kelvin set to 4000 (I don't want the light to get any cooler), the lights are too warm at 6:30am compared to the sunlight. Ideally, I'd like them to either match or be cooler than sunlight for those murky winter days to come. From what I see, customizing the shape of the color and brightness curves would allow me to do just that.

Yeah I see the same issue, but just changing the shape does not help, as the brightness of the natural light is vastly different, due to the low sun hitting the rooms through the windows much more and obstructions blocking the sun etc. 


I'm instead using a color temperature and brightness sensor in the rooms to determine how bright the light actually is and which color temperature mix it ends up in the end. 

This allows to steer the lights to the right brightness and temperature, to follow the adaptive lighting curve or an outdoor sensor. 

This leads to my lights changing a lot in their color temperature over the day, but you can see here, that it overall stays a lot warmer than it would be if adaptive lighting would run them.

The target for the room temperature and brightness was the outdoor conditions here. So the upper graph shows the light output and the lower graph shows the target from outside.

![Screenshot_2025-02-24-21-59-59-670-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/beb2b7a1-f052-4f42-aed3-8ef1614e5322)

And here for comparison what the indoor sensor reads (nearly a perfect match for the outdoor conditions):

![Screenshot_2025-02-24-22-03-24-921-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/e9cf923d-f54e-4d75-acfc-55f3f53341bc)
